### ISSUE TITLE
Accessing non-existent dataset from a thread produces errors on stdout

### ISSUE THREAD

**Issue by stuarteberg** [h5py/h5py#580]
Many Python programmers are accustomed to the ["EAFP"][1] idiom for reading (possibly non-existent) values from h5py files:

[1]: https://docs.python.org/2/glossary.html#term-eafp

```python
try:
    data = f[key].value
except KeyError:
    print "key doesn't exist:", key
```

h5py supports that idiom, but if the above code runs outside the main thread, the HDF5 library emits error messages to stdout.  (The error messages are disconcerting, but are otherwise harmless.)

Here's a simple test case:

```python
import h5py
import threading

def test():
    with h5py.File('/tmp/newfile.h5', 'w') as newfile:
        try:
            doesnt_exist = newfile['doesnt_exist'].value
        except KeyError:
            pass
    print "test complete."

th = threading.Thread(target=test)
th.start()
th.join()

print "hdf5:", h5py.version.hdf5_version
print "h5py:", h5py.__version__
print "DONE."
```

Which produces the following output on both Mac and Linux:

```
HDF5-DIAG: Error detected in HDF5 (1.8.14) thread 140017089001216:
  #000: H5O.c line 246 in H5Oopen(): unable to open object
    major: Symbol table
    minor: Can't open object
  #001: H5O.c line 1357 in H5O_open_name(): object not found
    major: Symbol table
    minor: Object not found
  #002: H5Gloc.c line 430 in H5G_loc_find(): can't find object
    major: Symbol table
    minor: Object not found
  #003: H5Gtraverse.c line 861 in H5G_traverse(): internal path traversal failed
    major: Symbol table
    minor: Object not found
  #004: H5Gtraverse.c line 641 in H5G_traverse_real(): traversal operator failed
    major: Symbol table
    minor: Callback failed
  #005: H5Gloc.c line 385 in H5G_loc_find_cb(): object 'doesnt_exist' doesn't exist
    major: Symbol table
    minor: Object not found
test complete.
hdf5: 1.8.14
h5py: 2.5.0
DONE.
```

I don't know enough about h5py internals to say for sure, but I wonder if this is related to an old issue that was fixed in h5py 2.1.1, originally discussed in the following thread:
https://groups.google.com/d/msg/h5py/fS5398h3cQI/fF8DCvWywPgJ

**Comment by andrewcollette** [comment]
Thanks!  I think the long-term fix for this may be to have api_gen.py simply disable the error stack printing for each call into HDF5.  Unfortunately there's not a way to turn this HDF5 "feature" off globally.

**Comment by stuarteberg** [comment]
This is also an issue for non-existent dataset attributes:

```python
import h5py
import threading
 
def test_attr():
    print "Starting test_attr."
    with h5py.File('/tmp/newfile.h5', 'w') as newfile:
        newfile['newdata'] = [1,2,3]
        try:
            nonexistent_attr = newfile['newdata'].attrs['nonexistent_attr']
        except KeyError:
            pass
    print "test_attr complete."
 
th = threading.Thread(target=test_attr)
th.start()
th.join()
 
print "hdf5:", h5py.version.hdf5_version
print "h5py:", h5py.__version__
print "DONE."
```

```
Starting test_attr.
HDF5-DIAG: Error detected in HDF5 (1.8.14) thread 140453902964480:
  #000: H5A.c line 476 in H5Aopen_by_name(): can't open attribute
    major: Attribute
    minor: Can't open object
  #001: H5Aint.c line 451 in H5A_open_by_name(): unable to load attribute info from object header
    major: Attribute
    minor: Unable to initialize object
  #002: H5Oattribute.c line 537 in H5O_attr_open_by_name(): can't locate attribute: 'nonexistent_attr'
    major: Attribute
    minor: Object not found
test_attr complete.
hdf5: 1.8.14
h5py: 2.5.0
DONE.
```

**Comment by stuarteberg** [comment]
For now, I'm using the following workaround in my code to avoid these errors.  I monkey-patch a few functions in h5py so that the errors are avoided.  It probably results in extra lookups, but for me, that's better than error output.

```python
def _monkey_patch_h5py():
    """
    This workaround avoids error messages from HDF5 when accessing non-existing
    files, datasets, and dataset attributes from non-main threads.

    See also:
    - https://github.com/h5py/h5py/issues/580
    - https://github.com/h5py/h5py/issues/582
    """
    import os
    import h5py

    old_dataset_getitem = h5py.Group.__getitem__
    def new_dataset_getitem(group, key):
        if key not in group:
            raise KeyError("Unable to open object (Object '{}' doesn't exist)".format( key ))
        return old_dataset_getitem(group, key)
    h5py.Group.__getitem__ = new_dataset_getitem

    old_file_init = h5py.File.__init__
    def new_file_init(f, name, mode=None, driver=None, libver=None, userblock_size=None, swmr=False, **kwds):
        if mode is None or mode == 'a':
            if not os.path.exists(name):
                mode = 'w'
        old_file_init(f, name, mode, driver, libver, userblock_size, swmr, **kwds)
    h5py.File.__init__ = new_file_init

    old_attr_getitem = h5py._hl.attrs.AttributeManager.__getitem__
    def new_attr_getitem(attrs, key):
        if key not in attrs:
            raise KeyError("Can't open attribute (Can't locate attribute: '{}')".format(key))
        return old_attr_getitem(attrs, key)
    h5py._hl.attrs.AttributeManager.__getitem__ = new_attr_getitem
```

**Comment by jreadey** [comment]
From hpy can you call: 
H5Eset_auto2(error_stack, NULL, NULL);
to turn off dumping errors to stderr?

**Comment by jreadey** [comment]
This is the page the discussing the above function: https://www.hdfgroup.org/HDF5/doc_resource/doc181/UG/13_ErrorHandling.html.

**Comment by stuarteberg** [comment]
I'm not familiar with the h5py internals, but that function sure looks promising...

**Comment by andrewcollette** [comment]
@jreadey, we do that already, here:

https://github.com/h5py/h5py/blob/master/h5py/_errors.pyx#L135

which is called in `__init__.py`when h5py starts up.  The problem is that it only applies to the current thread.

**Comment by jreadey** [comment]
Got it.  There's an older issue that discusses this: https://github.com/h5py/h5py/issues/206.

**Comment by andrewcollette** [comment]
@stuarteberg, do you have a moment to check if #583 works on your system?

**Comment by andrewcollette** [comment]
Fixed in #583.
